File: Game.py
import random
class Game:
    def __init__(self, player1_name):
        from Player import Player, Machine
        self.player1 = Player(player1_name)
        self.machine = Machine()
        self.current_player = self.player1
        self.opponent = self.machine
        self.turn = 1

    def start_game(self):
        for _ in range(5):
            self.player1.draw_card()
            self.machine.draw_card()
        self.who_plays_first()
        self.turn_loop()

    def turn_loop(self):
        while self.player1.life > 0 and self.machine.life > 0 and len(self.player1.deck) > 0 and len(self.machine.deck) > 0:
            print(f'TURNO: {self.turn}')
            if self.current_player == self.machine:
                print(f"\nTurno de {self.machine.name:<25}{'LP: ':>25}{self.machine.showLife()}")
                print("="*64)
                print(f"\n{self.current_player.name:<28} Campo de juego {self.opponent.name:>28}")
                
                print(self.opponent.display_field(True))
                print(self.current_player.display_field())
                self.play_machine_turn()
            else:
                print(f"\nTurno de {self.player1.name:<25}{'LP: ':>25}{self.current_player.showLife()}")
                print("="*64)
                print(f"\n{'':<25} Campo de juego {'':>25}")
                print(self.opponent.display_field(True))
                print(self.current_player.display_field())                
                self.play_turn()
                
            self.turn += 1
            self.switch_turn()
        self.declare_winner()

    def play_turn(self):
        from Card import MonsterCard, SpellCard, TrapCard
        from Position import Position
        self.current_player.draw_card()
        print(f"Mano de {self.current_player.name}:")
        self.current_player.show_hand()
        mano = self.current_player.hand
        input_continuar = 'S'
        monster_counter = 0
        while (any(isinstance(i, SpellCard) for i in mano) or any(isinstance(x, TrapCard) for x in mano) or monster_counter == 0) and input_continuar == 'S':
            input_continuar = input("Desea agregar una carta? S/N\n").upper()
            while input_continuar not in ['S', 'N']:
                input_continuar = input("Por favor, elige una opción válida (S/N): ").upper()
                
            if self.turn > 2 and  any(isinstance(c, MonsterCard) for c in self.current_player.field[0]):
                choice = input("Deseas cambiar la posición de alguna carta? (S/N): ")
                while choice not in ['S', 'N']:
                    choice = input("Por favor, elige una opción válida (S/N): ")
                    
                if choice == 'S':
                    card_index = input("Elige una carta para cambiar de posición: ")
                    while not card_index.isdigit() or int(card_index) >= len(self.current_player.field[0]):
                        card_index = input("Por favor, elige un número valido: ")
                    card_index = int(card_index)
                    card = self.current_player.field[0][card_index]
                    self.change_position(card)
            
            if input_continuar == 'S':        
                self.current_player.show_hand()
                choice = input("Elige una carta para jugar\n")
                while not choice.isdigit() or int(choice) >= len(self.current_player.hand):
                    choice = input(f"Por favor, elige un número valido (>= 0 | <= {len(self.current_player.hand)-1}): ")
                card_index = int(choice)
                card = self.current_player.hand[card_index]
                
                if isinstance(card, MonsterCard) and monster_counter == 0:
                    monster_counter += 1
                    card = self.current_player.play_card(card_index)
                    
                    positionChoice = input("Desea colocar la carta en posición de ataque o defensa? (A/D): ").upper()
                    while positionChoice not in ['A', 'D']:
                        positionChoice = input("Por favor, elige una posición válida (A/D): ").upper()
                        
                    if positionChoice == 'A':
                        card.setPosition(Position.FACE_UP_ATAQUE)
                    else:
                        card.setPosition(Position.FACE_DOWN)
                        
                    if self.turn > 2 and positionChoice == 'A':
                        attackChoice = input("Desea atacar con la carta? (S/N): ")
                        while attackChoice not in ['S', 'N']:
                            attackChoice = input("Por favor, elige una opción válida (S/N): ")
                        if attackChoice == 'S':
                            self.attack_phase(card)
                        elif attackChoice == 'N':
                            print("Perfecto, fin del turno")
                        
                elif card == None:
                    attack = input("Deseas atacar con otro monstruo en tu campo? (S/N): ")
                    while attack not in ['S', 'N']:
                        attack = input("Por favor, elige una posición válida (S/N): ")
                        
                    if attack == 'S':
                        self.current_player.show_hand()
                        monsterChoice = input("Perfecto elige un monstruo de tu campo con el cual deseas atacar")
                        while not monsterChoice.isdigit() or int(monsterChoice) >= len(self.current_player.field[0]):
                            monsterChoice = input(f"Por favor, elige un número valido (>= 0 | <= 3): ")
                            
                        card2 = self.current_player.field[0][int(monsterChoice)]
                        self.attack_phase(card2)
                        
                elif isinstance(card, MonsterCard) and monster_counter > 0:
                    print("No se puede colocar más de 1 monstruo por turno...")
                
                elif isinstance(card, SpellCard):
                    card = self.current_player.play_card(card_index)
                    for monster in self.current_player.field[0]:
                        if monster is not None:
                            isActived = card.activate(monster)
                            if isActived:
                                print(f"Carta {card.name} ha sido activada.")
                                self.current_player.remove_magic_card(card)
                
                elif isinstance(card, TrapCard):
                        card = self.current_player.play_card(card_index)

    def play_machine_turn(self):
        from Card import MonsterCard
        self.machine.draw_card()
        card = self.machine.playRandomCard(self.player1)
        if isinstance(card, MonsterCard) and self.turn > 2:
            self.attack_phase_machine(card)
        
    def change_position(self, monster_card):
        from Position import Position
        position = input("Desea cambiar la posición de la carta? (A/D): ")
        while position not in ['A', 'D']:
            position = input("Por favor, elige una posición válida (A/D): ")
        if position == 'A':
            monster_card.setPosition(Position.FACE_UP_ATAQUE)
        else:
            monster_card.setPosition(Position.FACE_DOWN)
            
    def attack_phase(self, monster_card):
        from Card import MonsterCard
        from Position import Position
        if isinstance(monster_card, MonsterCard):
            for index, card in enumerate(self.machine.field[0]):
                if card is not None and card.getPosition() == Position.FACE_UP_ATAQUE:
                    print(f'{index} - {card.name} | [{card.getAttack()}ATK]')
                elif card is not None and card.getPosition() == Position.FACE_UP_DEFENSA:
                    print(f'{index} - {card.name} | [{card.getDefense()}DEF]')
                elif card is not None and card.getPosition() == Position.FACE_DOWN:
                    print(f'{index} - Carta Boca Abajo')
                
            index_ataque = int(input(f"Elija un objetivo a atacar:"))
                
            while index_ataque >= 3:
                index_ataque = int(input(f"Por favor, elige un número valido (>= 0 | <= {len(self.machine.field[0])-1}): "))
                
            enemy_monster = self.machine.field[0][index_ataque]
            
            while enemy_monster is None and any(isinstance(i, MonsterCard) for i in self.machine.field[0]):
                index_ataque = int(input("Lo siento no puedes atacar a una posición vacía. Elige la posición de un monstruo enemigo para atacar: "))
                while index_ataque >= 3:
                    index_ataque = int(input(f"Por favor, elige un número valido (>= 0 | <= {len(self.machine.field[0])-1}): "))
                enemy_monster = self.machine.field[0][index_ataque]

            if not any(isinstance(i, MonsterCard) for i in self.machine.field[0]):
                print("No hay monstruos en el campo enemigo.")
                print(f"{monster_card.name} ataca directamente a {self.opponent.name}")
                self.opponent.take_damage(monster_card.getAttack())
            else:
                monster_card.perform_attack(enemy_monster, self.machine, self.player1)
    
    def attack_phase_machine(self, monster_card):
        print("Eligiendo un objetivo de ataque en el campo enemigo.")
        opponent_monster = self.opponent.getWeakestMonster()
        if not opponent_monster:
            print("No hay monstruos en el campo enemigo.")
            print(f"{monster_card.name} ataca directamente a {self.opponent.name}")
            self.opponent.take_damage(monster_card.attack)
        else:
            print(f"{monster_card.name} ataca a {opponent_monster.name}")
            monster_card.perform_attack(opponent_monster, self.player1, self.machine)

    
    def who_plays_first(self):
        print("Decidiendo quien juega primero...")
        if random.choice([True, False]):
            self.current_player = self.player1
            self.opponent = self.machine
            print(f"{self.player1.name} juega primero.")
        else:
            self.current_player = self.machine
            self.opponent = self.player1
            print(f"{self.machine.name} juega primero.")

    def switch_turn(self):
        self.current_player, self.opponent = self.opponent, self.current_player

    def declare_winner(self):
        if self.player1.life > 0:
            print(f"{self.player1.name} ha ganado el juego!")
        else:
            print(f"{self.machine.name} ha ganado el juego!")


game = Game("PLAYER1")
game.start_game()

File: Game.java
package mygame;

import java.util.Random;
import java.util.Scanner;

public class Game {
    private Player player1;
    private Machine machine;
    private Player currentPlayer;
    private Player opponent;
    private int turn;

    private static final Scanner sc = new Scanner(System.in);

    public Game(String player1Name) {
        this.player1 = new Player(player1Name);
        this.machine = new Machine();
        this.currentPlayer = this.player1;
        this.opponent = this.machine;
        this.turn = 1;
    }

    public void startGame() {
        for (int i = 0; i < 5; i++) {
            player1.drawCard();
            machine.drawCard();
        }
        whoPlaysFirst();
        turnLoop();
    }

    private void turnLoop() {
        while (player1.showLife() > 0 && machine.showLife() > 0
                && player1.deck.size() > 0 && machine.deck.size() > 0) {
            System.out.println("TURNO: " + turn);

            if (currentPlayer == machine) {
                System.out.println("\nTurno de " + String.format("%-25s", machine.name) + String.format("%25s", "LP: ") + machine.showLife());
                System.out.println("================================================================");
                System.out.println("\n" + String.format("%-28s", currentPlayer.name) + " Campo de juego " 
                                   + String.format("%28s", opponent.name));
                playMachineTurn();
            } else {
                System.out.println("\nTurno de " + String.format("%-25s", player1.name) + String.format("%25s", "LP: ") + currentPlayer.showLife());
                System.out.println("================================================================");
                System.out.println("\n" + String.format("%-25s", "") + " Campo de juego " 
                                   + String.format("%25s", ""));
                System.out.println(opponent.displayField(true));
                System.out.println(currentPlayer.displayField(false));
                playTurn();
            }
            turn++;
            switchTurn();
        }
        declareWinner();
    }

    private void playTurn() {
        currentPlayer.drawCard();
        System.out.println("Mano de " + currentPlayer.name + ":");
        currentPlayer.showHand();

        int monsterCounter = 0;
        String inputContinuar = "S";

        while ((hasSpellOrTrap(currentPlayer.hand) || monsterCounter == 0) && inputContinuar.equals("S")) {
            System.out.print("Desea agregar una carta? S/N\n");
            inputContinuar = sc.nextLine().toUpperCase();
            while (!inputContinuar.equals("S") && !inputContinuar.equals("N")) {
                System.out.print("Por favor, elige una opción válida (S/N): ");
                inputContinuar = sc.nextLine().toUpperCase();
            }

            if (turn > 2 && hasMonsterOnField(currentPlayer)) {
                System.out.print("Deseas cambiar la posición de alguna carta? (S/N): ");
                String choice = sc.nextLine().toUpperCase();
                while (!choice.equals("S") && !choice.equals("N")) {
                    System.out.print("Por favor, elige una opción válida (S/N): ");
                    choice = sc.nextLine().toUpperCase();
                }
                if (choice.equals("S")) {
                    System.out.print("Elige una carta para cambiar de posición: ");
                    String cardIndexStr = sc.nextLine();
                    while (!isValidIndex(cardIndexStr, currentPlayer.field[0].length)) {
                        System.out.print("Por favor, elige un número valido: ");
                        cardIndexStr = sc.nextLine();
                    }
                    int cardIndex = Integer.parseInt(cardIndexStr);

                    while (currentPlayer.field[0][cardIndex] == null) {
                        System.out.print("No puedes cambiar la posición de una posición vacía. Elige la posición de un monstruo en tu campo: ");
                        cardIndexStr = sc.nextLine();
                        while (!isValidIndex(cardIndexStr, currentPlayer.field[0].length)) {
                            System.out.print("Por favor, elige un número valido: ");
                            cardIndexStr = sc.nextLine();
                        }
                        cardIndex = Integer.parseInt(cardIndexStr);
                    }
                    Card card = currentPlayer.field[0][cardIndex];
                    if (card instanceof MonsterCard) {
                        changePosition((MonsterCard) card);
                    }
                }
            }

            if (inputContinuar.equals("S")) {
                currentPlayer.showHand();
                System.out.print("Elige una carta para jugar\n");
                String choiceStr = sc.nextLine();
                while (!isValidIndex(choiceStr, currentPlayer.hand.size())) {
                    System.out.print("Por favor, elige un número valido (>= 0 | <= " + (currentPlayer.hand.size() - 1) + "): ");
                    choiceStr = sc.nextLine();
                }
                int cardIndex = Integer.parseInt(choiceStr);
                Card card = currentPlayer.hand.get(cardIndex);

                if (card instanceof MonsterCard && monsterCounter == 0) {
                    monsterCounter++;
                    card = currentPlayer.playCard(cardIndex);
                    System.out.print("Desea colocar la carta en posición de ataque o defensa? (A/D): ");
                    String positionChoice = sc.nextLine().toUpperCase();
                    while (!positionChoice.equals("A") && !positionChoice.equals("D")) {
                        System.out.print("Por favor, elige una posición válida (A/D): ");
                        positionChoice = sc.nextLine().toUpperCase();
                    }
                    if (positionChoice.equals("A")) {
                        ((MonsterCard) card).setPosition(Position.FACE_UP_ATAQUE);
                    } else {
                        ((MonsterCard) card).setPosition(Position.FACE_DOWN);
                    }

                    System.out.println(opponent.displayField(true));
                    System.out.println(currentPlayer.displayField(false));

                    if (turn > 2 && positionChoice.equals("A")) {
                        System.out.print("Desea atacar con la carta? (S/N): ");
                        String attackChoice = sc.nextLine().toUpperCase();
                        while (!attackChoice.equals("S") && !attackChoice.equals("N")) {
                            System.out.print("Por favor, elige una opción válida (S/N): ");
                            attackChoice = sc.nextLine().toUpperCase();
                        }
                        if (attackChoice.equals("S")) {
                            attackPhase((MonsterCard) card);
                        } else {
                            System.out.println("Perfecto, fin del turno");
                        }
                    }
                } else if (card == null) {
                    System.out.print("Deseas atacar con otro monstruo en tu campo? (S/N): ");
                    String attack = sc.nextLine().toUpperCase();
                    while (!attack.equals("S") && !attack.equals("N")) {
                        System.out.print("Por favor, elige una posición válida (S/N): ");
                        attack = sc.nextLine().toUpperCase();
                    }
                    if (attack.equals("S")) {
                        System.out.print("Perfecto elige un monstruo de tu campo con el cual deseas atacar: ");
                        String monsterChoice = sc.nextLine();
                        while (!isValidIndex(monsterChoice, currentPlayer.field[0].length)) {
                            System.out.print("Por favor, elige un número valido (>= 0 | <= 3): ");
                            monsterChoice = sc.nextLine();
                        }
                        int mChoice = Integer.parseInt(monsterChoice);
                        if (currentPlayer.field[0][mChoice] == null) {
                            System.out.print("No puedes atacar con una posición vacía. ");
                        } else {
                            MonsterCard card2 = (MonsterCard) currentPlayer.field[0][mChoice];
                            attackPhase(card2);
                            System.out.println(opponent.displayField(true));
                            System.out.println(currentPlayer.displayField(false));
                        }
                    }
                } else if (card instanceof MonsterCard && monsterCounter > 0) {
                    System.out.println("No se puede colocar más de 1 monstruo por turno...");
                } else if (card instanceof SpellCard) {
                    card = currentPlayer.playCard(cardIndex);
                    System.out.println(opponent.displayField(true));
                    System.out.println(currentPlayer.displayField(false));
                    if (card instanceof SpellCard) {
                        SpellCard spell = (SpellCard) card;
                        for (int i = 0; i < 3; i++) {
                            Card monster = currentPlayer.field[0][i];
                            if (monster instanceof MonsterCard) {
                                boolean isActivated = spell.activate((MonsterCard) monster);
                                if (isActivated) {
                                    System.out.println("Carta " + spell.name + " ha sido activada.");
                                    currentPlayer.removeMagicCard(spell);
                                    break;
                                }
                            }
                        }
                    }
                } else if (card instanceof TrapCard) {
                    card = currentPlayer.playCard(cardIndex);
                    System.out.println(opponent.displayField(true));
                    System.out.println(currentPlayer.displayField(false));
                }
            }

            if (inputContinuar.equals("N") && turn > 2) {
                System.out.print("Desear atacar con algún monstruo? (S/N): ");
                String attackChoice = sc.nextLine().toUpperCase();
                while (!attackChoice.equals("S") && !attackChoice.equals("N")) {
                    System.out.print("Por favor, elige una opción válida (S/N): ");
                    attackChoice = sc.nextLine().toUpperCase();
                }
                if (attackChoice.equals("S")) {
                    System.out.print("Perfecto elige un monstruo de tu campo con el cual deseas atacar: ");
                    String monsterChoice = sc.nextLine();
                    while (!isValidIndex(monsterChoice, currentPlayer.field[0].length)) {
                        System.out.print("Por favor, elige un número valido (>= 0 | <= 3): ");
                        monsterChoice = sc.nextLine();
                    }
                    int mChoice = Integer.parseInt(monsterChoice);
                    while (currentPlayer.field[0][mChoice] == null) {
                        System.out.print("No puedes atacar con una posición vacía. Elige la posición de un monstruo en tu campo para atacar: ");
                        monsterChoice = sc.nextLine();
                        while (!isValidIndex(monsterChoice, currentPlayer.field[0].length)) {
                            System.out.print("Por favor, elige un número valido (>= 0 | <= 3): ");
                            monsterChoice = sc.nextLine();
                        }
                        mChoice = Integer.parseInt(monsterChoice);
                    }
                    MonsterCard card2 = (MonsterCard) currentPlayer.field[0][mChoice];
                    attackPhase(card2);
                    System.out.println(opponent.displayField(true));
                    System.out.println(currentPlayer.displayField(false));
                }
            }
        }
    }

    private void playMachineTurn() {
        machine.drawCard();
        Card card = machine.playRandomCard(player1);
        if (card instanceof MonsterCard && turn > 2) {
            attackPhaseMachine((MonsterCard) card);
        }
    }

    private void changePosition(MonsterCard monsterCard) {
        System.out.print("Desea cambiar la posición de la carta? (A/D): ");
        String position = sc.nextLine().toUpperCase();
        while (!position.equals("A") && !position.equals("D")) {
            System.out.print("Por favor, elige una posición válida (A/D): ");
            position = sc.nextLine().toUpperCase();
        }
        if (position.equals("A")) {
            monsterCard.setPosition(Position.FACE_UP_ATAQUE);
        } else {
            monsterCard.setPosition(Position.FACE_DOWN);
        }
    }

    private void attackPhase(MonsterCard monsterCard) {
        System.out.println("Elija un objetivo a atacar:");
        for (int i = 0; i < 3; i++) {
            Card c = machine.field[0][i];
            if (c instanceof MonsterCard) {
                MonsterCard mc = (MonsterCard) c;
                if (mc.getPositionValue() == Position.FACE_UP_ATAQUE.getValue()) {
                    System.out.println(i + " - " + mc.name + " | [" + mc.getAttack() + "ATK]");
                } else if (mc.getPositionValue() == Position.FACE_UP_DEFENSA.getValue()) {
                    System.out.println(i + " - " + mc.name + " | [" + mc.getDefense() + "DEF]");
                } else if (mc.getPositionValue() == Position.FACE_DOWN.getValue()) {
                    System.out.println(i + " - Carta Boca Abajo");
                }
            }
        }

        String indexStr = sc.nextLine();
        int indexAtaque = parseSafeInt(indexStr, -1);
        while (indexAtaque < 0 || indexAtaque >= 3) {
            System.out.print("Por favor, elige un número valido (>= 0 | <= 2): ");
            indexAtaque = parseSafeInt(sc.nextLine(), -1);
        }
        Card enemyMonster = machine.field[0][indexAtaque];

        if (!hasMonsters(machine.field[0])) {
            System.out.println("No hay monstruos en el campo enemigo.");
            System.out.println(monsterCard.name + " ataca directamente a " + opponent.name);
            opponent.takeDamage(monsterCard.getAttack());
            return;
        }

        while (enemyMonster == null && hasMonsters(machine.field[0])) {
            System.out.print("No puedes atacar una posición vacía mientras existan enemigos. Elige posición de un monstruo enemigo: ");
            indexAtaque = parseSafeInt(sc.nextLine(), -1);
            while (indexAtaque < 0 || indexAtaque >= 3) {
                System.out.print("Por favor, elige un número valido (>= 0 | <= 2): ");
                indexAtaque = parseSafeInt(sc.nextLine(), -1);
            }
            enemyMonster = machine.field[0][indexAtaque];
        }

        if (enemyMonster instanceof MonsterCard) {
            monsterCard.performAttack((MonsterCard) enemyMonster, machine, player1);
        }
    }

    private void attackPhaseMachine(MonsterCard machineMonsterCard) {
        System.out.println("Eligiendo un objetivo de ataque en el campo enemigo.");
        MonsterCard opponentMonster = player1.getWeakestMonster();
        if (opponentMonster == null) {
            System.out.println("No hay monstruos en el campo enemigo.");
            System.out.println(machineMonsterCard.name + " ataca directamente a " + opponent.name);
            opponent.takeDamage(machineMonsterCard.getAttack());
        } else {
            System.out.println(machineMonsterCard.name + " ataca a " + opponentMonster.name);
            machineMonsterCard.performAttack(opponentMonster, player1, machine);
        }
    }

    private void whoPlaysFirst() {
        System.out.println("Decidiendo quien juega primero...");
        if (new Random().nextBoolean()) {
            currentPlayer = player1;
            opponent = machine;
            System.out.println(player1.name + " juega primero.");
        } else {
            currentPlayer = machine;
            opponent = player1;
            System.out.println(machine.name + " juega primero.");
        }
    }

    private void switchTurn() {
        Player temp = currentPlayer;
        currentPlayer = opponent;
        opponent = temp;
    }

    private void declareWinner() {
        if (player1.showLife() > 0) {
            System.out.println(player1.name + " ha ganado el juego!");
        } else {
            System.out.println(machine.name + " ha ganado el juego!");
        }
    }

    private boolean isValidIndex(String str, int size) {
        try {
            int val = Integer.parseInt(str);
            return val >= 0 && val < size;
        } catch (NumberFormatException e) {
            return false;
        }
    }

    private int parseSafeInt(String s, int def) {
        try {
            return Integer.parseInt(s);
        } catch (NumberFormatException e) {
            return def;
        }
    }

    private boolean hasSpellOrTrap(java.util.List<Card> handList) {
        for (Card c : handList) {
            if (c instanceof SpellCard || c instanceof TrapCard) {
                return true;
            }
        }
        return false;
    }

    private boolean hasMonsterOnField(Player p) {
        for (int i = 0; i < 3; i++) {
            if (p.field[0][i] instanceof MonsterCard) {
                return true;
            }
        }
        return false;
    }

    private boolean hasMonsters(Card[] row) {
        for (Card c : row) {
            if (c instanceof MonsterCard) {
                return true;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        Game game = new Game("PLAYER1");
        game.startGame();
    }
}